File: opennotes-discord/src/utils/platform-claims.ts
import jwt from 'jsonwebtoken';
import { config } from '../config.js';
import { logger } from '../logger.js';

const PLATFORM_CLAIMS_EXPIRY_SECONDS = 300;
const JWT_ALGORITHM = 'HS256';

export interface PlatformClaims {
  platform: string;
  scope: string;
  sub: string;
  community_id: string;
  can_administer_community: boolean;
  type: 'platform_claims';
  iat: number;
  exp: number;
}

export function createPlatformClaimsToken(
  platform: string,
  scope: string,
  userId: string,
  communityId: string,
  canAdministerCommunity: boolean
): string | null {
  if (!config.jwtSecretKey) {
    logger.warn('JWT_SECRET_KEY not configured, cannot create platform claims token');
    return null;
  }

  const now = Math.floor(Date.now() / 1000);
  const exp = now + PLATFORM_CLAIMS_EXPIRY_SECONDS;

  const payload: PlatformClaims = {
    platform,
    scope,
    sub: userId,
    community_id: communityId,
    can_administer_community: canAdministerCommunity,
    type: 'platform_claims',
    iat: now,
    exp: exp,
  };

  return jwt.sign(payload, config.jwtSecretKey, {
    algorithm: JWT_ALGORITHM,
  });
}

File: opennotes-discourse/plugin/lib/opennotes/platform_claims.rb
# frozen_string_literal: true

module OpenNotes
  module PlatformClaims
    ALGORITHM = "HS256"
    TOKEN_LIFETIME = 3600

    def self.claims_for(user)
      {
        platform: "discourse",
        scope: Discourse.current_hostname,
        sub: user.id.to_s,
        username: user.username,
        trust_level: user.trust_level,
        admin: user.admin?,
        moderator: user.moderator?,
      }
    end

    def self.sign(user:, scope: nil, secret:)
      payload = claims_for(user)
      payload[:scope] = scope if scope
      now = Time.now.to_i
      payload[:iat] = now
      payload[:exp] = now + TOKEN_LIFETIME
      JWT.encode(payload, secret, ALGORITHM)
    end
  end
end
